File: vi/codes/cpp/chapter_hashing/array_hash_map.cpp
/**
 * File: array_hash_map.cpp
 * Created Time: 2022-12-14
 * Author: msk397 (machangxinq@gmail.com)
 */

#include "../utils/common.hpp"

/* Cặp khóa-giá trị */
struct Pair {
  public:
    int key;
    string val;
    Pair(int key, string val) {
        this->key = key;
        this->val = val;
    }
};

/* Bảng băm cài đặt dựa trên mảng */
class ArrayHashMap {
  private:
    vector<Pair *> buckets;

  public:
    ArrayHashMap() {
        // Khởi tạo mảng với 100 bucket (ngăn chứa)
        buckets = vector<Pair *>(100);
    }

    ~ArrayHashMap() {
        // Giải phóng bộ nhớ
        for (const auto &bucket : buckets) {
            delete bucket;
        }
        buckets.clear();
    }

    /* Hàm băm */
    int hashFunc(int key) {
        int index = key % 100;
        return index;
    }

    /* Thao tác truy vấn */
    string get(int key) {
        int index = hashFunc(key);
        Pair *pair = buckets[index];
        if (pair == nullptr)
            return "";
        return pair->val;
    }

    /* Thao tác thêm */
    void put(int key, string val) {
        Pair *pair = new Pair(key, val);
        int index = hashFunc(key);
        buckets[index] = pair;
    }

    /* Thao tác xóa */
    void remove(int key) {
        int index = hashFunc(key);
        // Giải phóng bộ nhớ và gán về nullptr
        delete buckets[index];
        buckets[index] = nullptr;
    }

    /* Lấy tất cả các cặp khóa-giá trị */
    vector<Pair *> pairSet() {
        vector<Pair *> pairSet;
        for (Pair *pair : buckets) {
            if (pair != nullptr) {
                pairSet.push_back(pair);
            }
        }
        return pairSet;
    }

    /* Lấy tất cả các khóa */
    vector<int> keySet() {
        vector<int> keySet;
        for (Pair *pair : buckets) {
            if (pair != nullptr) {
                keySet.push_back(pair->key);
            }
        }
        return keySet;
    }

    /* Lấy tất cả các giá trị */
    vector<string> valueSet() {
        vector<string> valueSet;
        for (Pair *pair : buckets) {
            if (pair != nullptr) {
                valueSet.push_back(pair->val);
            }
        }
        return valueSet;
    }

    /* In bảng băm */
    void print() {
        for (Pair *kv : pairSet()) {
            cout << kv->key << " -> " << kv->val << endl;
        }
    }
};

// Xem array_hash_map_test.cpp để biết các trường hợp kiểm thử

File: vi/codes/go/chapter_hashing/array_hash_map.go
// File: array_hash_map.go
// Created Time: 2022-12-14
// Author: msk397 (machangxinq@gmail.com)

package chapter_hashing

import "fmt"

/* Cặp khóa-giá trị */
type pair struct {
	key int
	val string
}

/* Bảng băm cài đặt dựa trên mảng */
type arrayHashMap struct {
	buckets []*pair
}

/* Khởi tạo bảng băm */
func newArrayHashMap() *arrayHashMap {
	// Khởi tạo mảng với 100 bucket
	buckets := make([]*pair, 100)
	return &arrayHashMap{buckets: buckets}
}

/* Hàm băm */
func (a *arrayHashMap) hashFunc(key int) int {
	index := key % 100
	return index
}

/* Thao tác truy vấn */
func (a *arrayHashMap) get(key int) string {
	index := a.hashFunc(key)
	pair := a.buckets[index]
	if pair == nil {
		return "Not Found"
	}
	return pair.val
}

/* Thao tác thêm */
func (a *arrayHashMap) put(key int, val string) {
	pair := &pair{key: key, val: val}
	index := a.hashFunc(key)
	a.buckets[index] = pair
}

/* Thao tác xóa */
func (a *arrayHashMap) remove(key int) {
	index := a.hashFunc(key)
	// Đặt thành nil để xóa
	a.buckets[index] = nil
}

/* Lấy tất cả cặp khóa-giá trị */
func (a *arrayHashMap) pairSet() []*pair {
	var pairs []*pair
	for _, pair := range a.buckets {
		if pair != nil {
			pairs = append(pairs, pair)
		}
	}
	return pairs
}

/* Lấy tất cả khóa */
func (a *arrayHashMap) keySet() []int {
	var keys []int
	for _, pair := range a.buckets {
		if pair != nil {
			keys = append(keys, pair.key)
		}
	}
	return keys
}

/* Lấy tất cả giá trị */
func (a *arrayHashMap) valueSet() []string {
	var values []string
	for _, pair := range a.buckets {
		if pair != nil {
			values = append(values, pair.val)
		}
	}
	return values
}

/* In bảng băm */
func (a *arrayHashMap) print() {
	for _, pair := range a.buckets {
		if pair != nil {
			fmt.Println(pair.key, "->", pair.val)
		}
	}
}
